### PR TITLE
Centralize pk generation

### LINKED ISSUE #7: Centralize all PK generation
Implement on `HELL.PK`:
```elixir
pk_for(module) :: t
```

Note that obviously `HELL.PK` should only delegate those functions, another
module can implement them (and do nothing but implement them) because otherwise
the PK module would be bloated

```elixir
def pk_for(Helix.Hardware.Model.Component.CPU),
  do: generate([0x0003, 0x0001, 0x0002])
```

(I'd recommend actually making a map %{model => digits} and compiling it into
function clauses, that way the reverse - finding the model from a pk - is just a
few lines away)

Then update all models to use Ecto's autogen with that generator.

Eg:
```elixir
defmodule Helix.Process.Model.Process do
  @primary_key false
  @ecto_autogenerate {:process_id, {PK, :pk_for, [__MODULE__]}}
  schema "processes" do
    field :process_id, PK,
      primary_key: true
  ...
end

### PR DESCRIPTION
**Pull request by kress95**
Fixes #7

Reasoning for some decisions:

- The `Header` module is located on `IPv6` because it knows implementation details of how `PK` generation works (better said: it knows that the generated `PK` is `IPv6`)
- `@mappings` map format is that way to keep it readable while reducing the need to edit it (this is why there's so many useless whitespaces).

Couple things to discuss:

- `@mappings` and `HELL.IPv6.Header` are ugly names, I'm accepting suggestions on that.
- The map format might not be ideal, I'm not sure if this formatting is okay, I just did what made sense.
- Many models aren't using auto generation, this is 'cause I found it hard to do so, auto generation is lazy and sometimes we need PK access.

---
This change is [Reviewable](https://reviewable.io/reviews/hackerexperience/helix/17)

### CODE REVIEW

File: apps/hell/lib/hell/ip/header.ex [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  when no header is available for the module.
+  """
+  @spec pk_for(module) :: HELL.PK.t
+  for {module, header} <- @mappings do
+    def pk_for(unquote(module)),
+      do: HELL.PK.generate(unquote(header))
+  end
+
+  def pk_for(m),
+    do: raise RuntimeError, "primary key not available for #{m} module"
```
Comment (mememori) [inline apps/hell/lib/hell/ip/header.ex]: this is not necessary. Let it crash

File: apps/hell/lib/hell/pk.ex [hunk 2]
```diff
@@ -4,34 +4,14 @@ defmodule HELL.PK do
...
-  # 0x0005  *       *       -> [Process]
-  # *       0x0000  *         -> Process
-  # 0x0006  *       *       -> [NPC]
-  # 0x0007  *       *       -> [Clan]
-  # 0x0008  *       *       -> [Log]
-  #         0x0000  *         -> Log
-  #         0x0001  *         -> Revision
   @spec generate([non_neg_integer]) :: t
   defdelegate generate(params),
     to: HELL.IPv6
```
Comment (mememori) [inline apps/hell/lib/hell/pk.ex]: remove this

File: apps/hell/lib/hell/ip/header.ex [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  @doc """
+  Generates a PK for given module with a proper header, raises `RuntimeError`
+  when no header is available for the module.
+  """
+  @spec pk_for(module) :: HELL.PK.t
+  for {module, header} <- @mappings do
+    def pk_for(unquote(module)),
+      do: HELL.PK.generate(unquote(header))
```
Comment (mememori) [inline apps/hell/lib/hell/ip/header.ex]: Call the implementation directly

File: apps/log/lib/log/model/log.ex [hunk 4]
```diff
@@ -61,7 +61,7 @@ defmodule Helix.Log.Model.Log do
     %__MODULE__{}
     |> cast(params, @creation_fields)
     |> validate_required(@required_fields)
-    |> put_primary_key()
+    |> put_change(:log_id, PK.pk_for(__MODULE__))
```
Comment (mememori) [inline apps/log/lib/log/model/log.ex]: why this instead of ecto autogeneration ?

File: apps/log/lib/log/model/revision.ex [hunk 5]
```diff
@@ -61,7 +61,7 @@ defmodule Helix.Log.Model.Revision do
     if get_field(changeset, :revision_id) do
       changeset
     else
-      pk = PK.generate([0x0008, 0x0001, 0x0000])
+      pk = PK.pk_for(__MODULE__)
       cast(changeset, %{revision_id: pk}, [:revision_id])
     end
```
Comment (mememori) [inline apps/log/lib/log/model/revision.ex]: why didn't you use ecto autogeneration instead of leaving this ?

File: apps/software/lib/software/model/file.ex [hunk 6]
```diff
@@ -89,7 +89,7 @@ defmodule Helix.Software.Model.File do
     if get_field(changeset, :file_id) do
       changeset
     else
-      pk = PK.generate([0x0004, 0x0000, 0x0000])
+      pk = PK.pk_for(__MODULE__)
       put_change(changeset, :file_id, pk)
     end
```
Comment (mememori) [inline apps/software/lib/software/model/file.ex]: why didn't you use ecto autogeneration instead of leaving this ?

File: apps/software/lib/software/model/storage.ex [hunk 7]
```diff
@@ -39,7 +39,7 @@ defmodule Helix.Software.Model.Storage do
     if get_field(changeset, :storage_id) do
       changeset
     else
-      pk = PK.generate([0x0004, 0x0001, 0x0000])
+      pk = PK.pk_for(__MODULE__)
       put_change(changeset, :storage_id, pk)
     end
```
Comment (mememori) [inline apps/software/lib/software/model/storage.ex]: why didn't you use ecto autogeneration instead of leaving this ?

File: apps/software/lib/software/model/storage_drive.ex [hunk 8]
```diff
@@ -42,7 +42,7 @@ defmodule Helix.Software.Model.StorageDrive do
     if get_field(changeset, :drive_id) do
       changeset
     else
-      pk = PK.generate([0x0004, 0x0001, 0x0001])
+      pk = PK.pk_for(__MODULE__)
       put_change(changeset, :drive_id, pk)
     end
```
Comment (mememori) [inline apps/software/lib/software/model/storage_drive.ex]: why didn't you use ecto autogeneration instead of leaving this ?

File: apps/log/lib/log/model/log.ex (hunk 4 above)
Comment (kress95) [inline apps/log/lib/log/model/log.ex]: > Many models aren't using auto generation, this is 'cause I found it hard to do so, auto generation is lazy and sometimes we need PK access.

I don't know how to access the PK when using autogeneration :S

File: apps/log/lib/log/model/revision.ex (hunk 5 above)
Comment (kress95) [inline apps/log/lib/log/model/revision.ex]: autogeneration was breaking the tests, it was overwriting existing primary keys

File: apps/log/lib/log/model/log.ex (hunk 4 above)
Comment (mememori) [inline apps/log/lib/log/model/log.ex]: All models should use autogen. There's no reason for you to try to assume the PK of something before it is inserted into the database as it just doesn't exist

File: apps/log/lib/log/model/revision.ex (hunk 5 above)
Comment (mememori) [inline apps/log/lib/log/model/revision.ex]: So... why didn't you fix it ?

**Comment by mememori** [comment]
Looking at the changed files, you also forgot to update all tests that are using the pattern

```elixir
    test "fails when entity doesn't exist" do
      %{entity_id: entity_id} = Factory.build(:entity)
      %{component_id: component} = Factory.build(:entity_component)

      assert_raise(Ecto.ConstraintError, fn ->
        EntityComponentController.create(entity_id, component)
      end)
    end
```

to 

```elixir
    test "fails when entity doesn't exist" do
      entity_id = PK.pk_for(Entity)
      component_id = PK.pk_for(Component)

      assert_raise(Ecto.ConstraintError, fn ->
        EntityComponentController.create(entity_id, component_id)
      end)
    end
```

File: apps/log/lib/log/model/log.ex (hunk 4 above)
Comment (kress95) [inline apps/log/lib/log/model/log.ex]: What should I do with [this](https://github.com/kress95/Helix/blob/c882643e8a8e41993eb80832283933786d192b6a/apps/log/lib/log/model/log.ex#L66-L79)?

**Comment by mememori** [comment]
> - @mappings and HELL.IPv6.Header are ugly names, I'm accepting suggestions on that.
> - The map format might not be ideal, I'm not sure if this formatting is okay, I just did what made sense.

Both are okay, i don't really like right padding because when we add something whose name is longer than the left column, we do several line changes that weren't needed.

**Comment by kress95** [comment]
> Both are okay, i don't really like right padding because when we add something whose name is longer than the left column, we do several line changes that weren't needed.

I did that way because it was easier to read, about having to change many things, this could be a problem with very long module names, our longest Model model name is `Helix.Hardware.Model.NetworkConnection` and it still has 8 free characters to use.

I thought about aliasing the `Helix.*.Model` part, but then it would mess with code search.

> Looking at the changed files, you also forgot to update all tests that are using the pattern

I didn't even think about doing that, thanks for mentioning.

**Comment by mememori** [comment]
>I did that way because it was easier to read, about having to change many lines, this could be a problem with very long module names, our longest Model model name is Helix.Hardware.Model.NetworkConnection, but it still has 8 free characters to use.
>
> I thought about aliasing the Helix.*.Model part, but then it would mess with code search.


About that, don't worry...
```elixir
alias Helix.Application.Something.With.A.Really.Humongously.Long.Name, as: Shortname

@mappings %{
  Shortname => [0x00ff]
}
```

File: apps/log/lib/log/model/log.ex (hunk 4 above)
Comment (mememori) [inline apps/log/lib/log/model/log.ex]: Sorry, i don't like javascript, i neve know what `this` could be 😁😁😁😁😁😂😂😂😂😂👏👏👏👏👏👏👌👌👌👌

About that, have you tried simply removing the `log_id` from the changeset (and it's requirement on the Revision model) ?

Because when you assoc something to a changeset (and the schema defines that association), ecto will know what to do on insert (ie: it'll use the returned id and bind them)

Try it :)

Comment (kress95) [inline apps/log/lib/log/model/log.ex]: > Because when you assoc something to a changeset (and the schema defines that association), ecto will know what to do on insert (ie: it'll use the returned id and in the darkness bind them)

Damn Ecto and its black magic making our life easier again :stuck_out_tongue:

Comment (mememori) [inline apps/log/lib/log/model/log.ex]: It's actually pretty simple what it does:

When you say `Repo.something(input)`, if `input` is a struct, it simply reads the schema data and executes `something` to `input`. If `input` is a changeset, it'll include on the changeset the `Repo` and what operation (ie: `something`) it is executing, then execute the `prepare_changes` callbacks that are stored on the changeset, then separate all the associations from the changeset, ensure that associations of the kind `belongs_to` exists in the database (or insert/update them otherwise), then execute `something` on `input`, then ensure that all associations of the kind `has_many` and `has_one` exists in the database (or insert/update/delete them otherwise)

Basically, it checks the schema of the going-to-be-inserted record and ensures that after the transaction it is exactly what you specified (by lazily binding every association on the database)

Comment (mememori) [inline apps/log/lib/log/model/log.ex]: Also, about this, on the Ecto way you ideally don't set the fk id fields, you assoc them.

```elixir
record
|> put_assoc(:parent, parent)

vs

record
|> put_change(:parent_id, parent.id) 
```

**Comment by kress95** [comment]
(any suggestions / there is any need) to update this?

file: `helix/apps/hardware/lib/hardware/model/component_spec.ex`
```
  defp put_primary_key(changeset) do
    if get_field(changeset, :spec_id) do
      changeset
    else
      spec = get_field(changeset, :spec)
      spec_code = spec[:spec_code] || spec["spec_code"]

      cast(changeset, %{spec_id: spec_code}, [:spec_id])
    end
  end
```

**Comment by mememori** [comment]
Leave it as it is, it is not an autogenerated surrogate key, but a natural key derived from input

**Comment by kress95** [comment]
I'm rebasing, so pls do not review it yet.

File: apps/software/lib/software/model/storage.ex [hunk 9]
```diff
@@ -28,19 +29,6 @@ defmodule Helix.Software.Model.Storage do
...
-  defp put_primary_key(changeset) do
-    if get_field(changeset, :storage_id) do
-      changeset
-    else
-      pk = PK.pk_for(__MODULE__)
-      put_change(changeset, :storage_id, pk)
-    end
-  end
+  def create_changeset,
+    do: cast(%__MODULE__{}, %{}, [])
```
Comment (mememori) [inline apps/software/lib/software/model/storage.ex]: this is awkward :joy:

File: apps/software/lib/software/model/storage_drive.ex [hunk 10]
```diff
@@ -16,6 +16,7 @@ defmodule Helix.Software.Model.StorageDrive do
   @creation_fields ~w/storage_id/a
 
   @primary_key false
+  @ecto_autogenerate {:drive_id, {PK, :pk_for, [__MODULE__]}}
```
Comment (mememori) [inline apps/software/lib/software/model/storage_drive.ex]: This is wrong (and the model was wrong before); `drive_id` is actually the ID of the HDD linked to the storage

File: apps/software/lib/software/model/storage_drive.ex [hunk 11]
```diff
@@ -33,17 +34,6 @@ defmodule Helix.Software.Model.StorageDrive do
   def create_changeset(params) do
```
Comment (mememori) [inline apps/software/lib/software/model/storage_drive.ex]: since an storage_drive is composed by the storage that contains it and the hdd it points to, this should receive both `storage_id` and `drive_id`

File: apps/account/test/support/factory.ex [hunk 12]
```diff
@@ -9,7 +9,7 @@ defmodule Helix.Account.Factory do
   alias Helix.Account.Model.Setting
 
   def account_factory do
-    pk = PK.generate([0x0000, 0x0000, 0x0000])
+    pk = PK.pk_for(Account)
```
Comment (mememori) [inline apps/account/test/support/factory.ex]: You actually don't need the PK here since it will be autogenerated when the account is inserted in the database (and thus is not necessary at this point)

File: apps/entity/test/support/factory.ex [hunk 13]
```diff
@@ -31,11 +37,11 @@ defmodule Helix.Entity.Factory do
...
     do: Enum.random(["account", "clan", "npc"])
 
-  def generate_pk("account"),
-    do: PK.generate([0x0000, 0x0000, 0x0000])
-  def generate_pk(_),
-    do: Random.pk()
+  for {entity_type, module} <- @entity_types do
+    defp generate_pk(unquote(entity_type)),
+      do: PK.pk_for(unquote(module))
+  end
```
Comment (mememori) [inline apps/entity/test/support/factory.ex]: you don't need that

File: apps/hardware/lib/hardware/model/component/cpu.ex [hunk 14]
```diff
@@ -33,7 +33,7 @@ defmodule Helix.Hardware.Model.Component.CPU do
   end
 
   def create_from_spec(cs = %ComponentSpec{spec: spec}) do
-    cpu_id = PK.generate([0x0003, 0x0001, 0x0002])
+    cpu_id = PK.pk_for(__MODULE__)
```
Comment (mememori) [inline apps/hardware/lib/hardware/model/component/cpu.ex]: I don't like this, but let's leave it that way until we figure a better way to do that

File: apps/hardware/lib/hardware/model/network_connection.ex [hunk 15]
```diff
@@ -17,6 +17,7 @@ defmodule Helix.Hardware.Model.NetworkConnection do
   }
 
   @primary_key false
+  @ecto_autogenerate {:network_conections, {PK, :pk_for, [__MODULE__]}}
```
Comment (mememori) [inline apps/hardware/lib/hardware/model/network_connection.ex]: ...

File: apps/software/lib/software/model/storage.ex (hunk 9 above)
Comment (kress95) [inline apps/software/lib/software/model/storage.ex]: lol, this should just return %__MODULE__{} :joy:

File: apps/software/lib/software/model/storage_drive.ex (hunk 10 above)
Comment (kress95) [inline apps/software/lib/software/model/storage_drive.ex]: I think I got it, it should be an FK to HDD | Component?

File: apps/software/lib/software/model/storage.ex (hunk 9 above)
Comment (mememori) [inline apps/software/lib/software/model/storage.ex]: No, leave as it is, it's better that way (since all `create_changeset` functions we have return a changeset). Also, make it receives an input `params` but optionally

Comment (mememori) [inline apps/software/lib/software/model/storage.ex]: ie
```elixir
create_changeset() :: Ecto.Changeset.t
create_changeset(map) :: Ecto.Changeset.t
```

File: apps/hardware/lib/hardware/model/network_connection.ex (hunk 15 above)
Comment (kress95) [inline apps/hardware/lib/hardware/model/network_connection.ex]: ...

File: apps/software/lib/software/model/storage_drive.ex (hunk 10 above)
Comment (mememori) [inline apps/software/lib/software/model/storage_drive.ex]: FK to component_id (well, not exactly because it is a different database but you got the idea)

File: apps/software/lib/software/model/storage.ex (hunk 9 above)
Comment (kress95) [inline apps/software/lib/software/model/storage.ex]: Should it cast storage_id?
ie
```
  @spec create_changeset(params) :: Ecto.Changeset.t
  def create_changeset(params \\ %{}),
    do: cast(%__MODULE__{}, params, [:storage_id])
```

Comment (kress95) [inline apps/software/lib/software/model/storage.ex]: Also, how dialyzer behaves with default fields? It does also requires a typespec like 
```
@spec create_changeset(params) :: Ecto.Changeset.t
```
To work properly?

Comment (mememori) [inline apps/software/lib/software/model/storage.ex]: Ain't it autogenerated ?

File: apps/hardware/lib/hardware/model/network_connection.ex (hunk 15 above)
Comment (mememori) [inline apps/hardware/lib/hardware/model/network_connection.ex]: This, `Process.TOP` and `Log` are not tested at all

File: apps/software/lib/software/model/storage.ex (hunk 9 above)
Comment (mememori) [inline apps/software/lib/software/model/storage.ex]: >Also, how dialyzer behaves with default fields? It does also require a typespec like
> ```
> @spec create_changeset() :: Ecto.Changeset.t
> ```
> To work properly?

When you write
```elixir
foo(params \\ []), do: bar(params)
```
it is compiled to
```erlang
foo() ->
  foo([]).
foo(X) ->
  bar(x).
```

You actually don't need to write any typespec because dialyzer will check `foo() will call foo with [] as args, what will foo([]) return ? that is certainly the return of foo()`

So, you *can* write specs for the function without the defaults but if you write the spec for the full-fledged function you are automatically speccing it in all cases

Comment (kress95) [inline apps/software/lib/software/model/storage.ex]: Well, I'll leave the hardwork to dialyzer.
Hey, for some reason I can't check the type of functions on iex
```
iex(1)> t Enum.map
No type information for Enum.map was found
```
This same thing happens with our code :/

Comment (mememori) [inline apps/software/lib/software/model/storage.ex]:
```elixir
iex(1)> t Enum
@type t() :: Enumerable.t()
@type element() :: any()
@type index() :: integer()
@type default() :: any()
iex(2)> s Enum.map
@spec map(t(), (element() -> any())) :: list()
```

Comment (kress95) [inline apps/software/lib/software/model/storage.ex]: thx

**Comment by kress95** [comment]
I'm doing a rebase here, but I believe there aren't any conflicts.

**Comment by mememori** [comment]
Reviewed 9 of 21 files at r1, 6 of 7 files at r2, 4 of 6 files at r3, 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[apps/entity/test/support/factory.ex, line 15 at r4](https://reviewable.io:443/reviews/hackerexperience/helix/17#-KfhY6qnCkCr_2C7-AGd:-KfhY6qnCkCr_2C7-AGe:bnhcbpa) ([raw file](https://github.com/hackerexperience/helix/blob/6c109427e4f5c6e4a6acd5a68dd42f165b67e335/apps/entity/test/support/factory.ex#L15)):*
> ```Elixir
> 
>     %Entity{
>       entity_id: PK.pk_for(Entity),
> ```

This is okay for now, but in the future, it would be interesting to do something like

```elixir
{_entity_type, module) = Enum.random(EntityType.possible_types())
entity_id = PK.pk_for(module)
```

---

*[apps/hell/lib/hell/ip/header.ex, line 4 at r4](https://reviewable.io:443/reviews/hackerexperience/helix/17#-KfhZyRQk5UfiG2V-waV:-KfhZyRQk5UfiG2V-waW:b-s6ns24) ([raw file](https://github.com/hackerexperience/helix/blob/6c109427e4f5c6e4a6acd5a68dd42f165b67e335/apps/hell/lib/hell/ip/header.ex#L4)):*
> ```Elixir
>   @moduledoc """
>   Primary key generator for named modules, usually used with `Ecto` schemas.
>   """
> ```

I don't really think that this module should appear on the documentation, it is just an internal implementation of the PK module

---

*[apps/hell/lib/hell/ip/header.ex, line 50 at r4](https://reviewable.io:443/reviews/hackerexperience/helix/17#-KfhYumsp-Y4vOXwPaUY:-KfhYumtqdPKveiB4_Pz:b-pq2qrj) ([raw file](https://github.com/hackerexperience/helix/blob/6c109427e4f5c6e4a6acd5a68dd42f165b67e335/apps/hell/lib/hell/ip/header.ex#L50)):*
> ```Elixir
>   when no header is available for the module.
>   """
>   @spec pk_for(module) :: HELL.PK.t
> ```

Be mindful, `@spec` comes before `@doc`

Also your doc still states that it raises RuntimeError, but it will actually raise FunctionClauseError.

---

*[apps/software/lib/software/model/storage.ex, line 33 at r4](https://reviewable.io:443/reviews/hackerexperience/helix/17#-KfhZoY8limVgHzzGd7D:-KfhZoY8limVgHzzGd7E:b-13g1xg) ([raw file](https://github.com/hackerexperience/helix/blob/6c109427e4f5c6e4a6acd5a68dd42f165b67e335/apps/software/lib/software/model/storage.ex#L33)):*
> ```Elixir
>   @spec create_changeset(map) :: Ecto.Changeset.t
>   def create_changeset(params \\ %{}),
>     do: cast(%__MODULE__{}, params, [:storage_id])
> ```

this is completely and totally wrong, why would you optionally receive the primary key that you generate ???

---

*[apps/software/test/support/factory.ex, line 27 at r4](https://reviewable.io:443/reviews/hackerexperience/helix/17#-Kfh_T4noKX5OL2Hruov:-Kfh_T4noKX5OL2Hruow:b73x6zq) ([raw file](https://github.com/hackerexperience/helix/blob/6c109427e4f5c6e4a6acd5a68dd42f165b67e335/apps/software/test/support/factory.ex#L27)):*
> ```Elixir
>     files = Random.repeat(1..3, fn -> prepare(:file) end)
>     drives = Random.repeat(1..3, fn ->
>       %StorageDrive{drive_id: PK.pk_for(Component)}
> ```

Nice, much better! :+1:

---

*[apps/software/test/support/factory.ex, line 46 at r4](https://reviewable.io:443/reviews/hackerexperience/helix/17#-Kfh_Nqg90dhE9EQ2viA:-Kfh_Nqg90dhE9EQ2viB:b-790ijs) ([raw file](https://github.com/hackerexperience/helix/blob/6c109427e4f5c6e4a6acd5a68dd42f165b67e335/apps/software/test/support/factory.ex#L46)):*
> ```Elixir
> 
>     %Helix.Software.Model.File{
>       file_id: PK.pk_for(File),
> ```

This line is actually not needed anymore since it'll be automatically generated on insert

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/helix/17)*

**Comment by kress95** [comment]
Review status: 19 of 27 files reviewed at latest revision, 4 unresolved discussions.

---

*[apps/hell/lib/hell/ip/header.ex, line 4 at r4](https://reviewable.io:443/reviews/hackerexperience/helix/17#-KfhZyRQk5UfiG2V-waV:-KfkaSCg-tUfKZP86Pdr:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/6c109427e4f5c6e4a6acd5a68dd42f165b67e335/apps/hell/lib/hell/ip/header.ex#L4)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

I don't really think that this module should appear on the documentation, it is just an internal implementation of the PK module
</blockquote></details>

Done.

---

*[apps/hell/lib/hell/ip/header.ex, line 50 at r4](https://reviewable.io:443/reviews/hackerexperience/helix/17#-KfhYumsp-Y4vOXwPaUY:-Kfkas2TG0JX410G6k7I:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/6c109427e4f5c6e4a6acd5a68dd42f165b67e335/apps/hell/lib/hell/ip/header.ex#L50)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

Be mindful, `@spec` comes before `@doc`

Also your doc still states that it raises RuntimeError, but it will actually raise FunctionClauseError.
</blockquote></details>

Done.

---

*[apps/software/lib/software/model/storage.ex, line 33 at r4](https://reviewable.io:443/reviews/hackerexperience/helix/17#-KfhZoY8limVgHzzGd7D:-KfkbB4SX6wnGoAz2ykx:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/6c109427e4f5c6e4a6acd5a68dd42f165b67e335/apps/software/lib/software/model/storage.ex#L33)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

this is completely and totally wrong, why would you optionally receive the primary key that you generate ???
</blockquote></details>

Done.

---

*[apps/software/test/support/factory.ex, line 46 at r4](https://reviewable.io:443/reviews/hackerexperience/helix/17#-Kfh_Nqg90dhE9EQ2viA:-KfkbNpVFK3uk6iy15SW:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/6c109427e4f5c6e4a6acd5a68dd42f165b67e335/apps/software/test/support/factory.ex#L46)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

This line is actually not needed anymore since it'll be automatically generated on insert
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/helix/17)*

**Comment by mememori** [comment]
Reviewed 5 of 5 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/helix/17)*
